src/streams.c

```diff
@@ -396,9 +396,11 @@ static void READ_TEST_OR_LOOP(Obj context)
 
             /* print the result                                            */
             if ( ! dualSemicolon ) {
+#ifdef GAP_KERNEL_DEBUG
                 Bag currLVars = STATE(CurrLVars); /* in case view runs into error */
+#endif
                 ViewObjHandler( evalResult );
-                SWITCH_TO_OLD_LVARS(currLVars);
+                GAP_ASSERT(currLVars == STATE(CurrLVars));
```

Comment:
so, I am still worried about this... There is similar code in `CALL_WITH_CATCH`. If I modify that in a similar way, I *can* trigger the assertion:
```
gap> CALL_WITH_CATCH(x->x,[1..3]);
Error, Function: number of arguments must be 1 (not 3)
not in any function at *stdin*:1
type 'quit;' to quit to outer loop
brk>
Assertion failed: (currLVars == (ActiveGAPState()->CurrLVars)), function CALL_WITH_CATCH, file ../../src/error.c, line 27
...
(lldb) p currLVars
(Obj) $0 = 0x0000100000010fe8
(lldb) p MainGAPState.CurrLVars
(Bag) $1 = 0x00001000005d1f28
(lldb) p PARENT_LVARS(MainGAPState.CurrLVars)
(Obj) $2 = 0x0000100000010fe8
```
So the active lvars at that point have the original ones as parent. I think this come from the break loop? Makes me wonder if `SHELL` should save & restore lvars... but are there other places this could happen?

Perhaps if a ViewObj method triggers an Error, could this assertion here be triggered? I *think* I tried that, but I should try again to be sure?

Hrm

Reply:
Aha. Turns out I tested this, but only on a break loop, not in a `.tst` file. And it turns out, I can trigger the assertion then: if I put this into a file `foo.tst`:
```
gap> InstallMethod(ViewObj, [IsStringRep], 100,
>     function(s)
>         if s="foobar" then Error("oops"); fi;
>         TryNextMethod();
>     end);
gap> "foobar";
Error, oops at *stdin*:1
```
Then I see this:
```
gap> Test("foo.tst");
# line 6 of 6 (100%)Assertion failed: (currLVars == (ActiveGAPState()->CurrLVars)), function READ_TEST_OR_LOOP, file ../../src/streams.c, line 403.
```

Reply:
So we must not merge this as-is. Now I tried and patched the kernel function `Shell` to perform the same assertion check, and see, and can trigger the assertion there with the same code snippet (now pasted into a REPL).

Need to analyze this further to see why not all hell breaks loose. Probably we are restoring CurrLVars elsewhere, but where? Need to work on other stuff now, though.